### PR TITLE
Implement Display, FromStr and Hash

### PR DESCRIPTION
**Pull request by kaidokert**
## Summary by Sourcery

Add string conversion support (Display and FromStr) and hashing for fixed-size big integers, fill in missing shift and hash implementations, update trait bounds, bump version, and cover new functionality with tests.

New Features:
- Implement Display formatter for FixedUInt<T, N>
- Implement FromStr parsing for FixedUInt<T, N>

Enhancements:
- Add signed_shl and signed_shr methods for FixedUInt PrimInt implementation
- Implement Hash trait for BytesHolder by hashing its internal array

Build:
- Bump crate version to 0.1.15

Tests:
- Add unit tests for Display and FromStr covering zero, small and large values, different types, round-trips, and error cases

Chores:
- Extend MachineWord trait bound to include core::hash::Hash

## Summary by CodeRabbit

* **New Features**
  * Added support for displaying and parsing fixed-size unsigned integers as decimal strings.
  * Implemented signed shift operations for fixed-size unsigned integers.

* **Bug Fixes**
  * Enabled correct hashing for byte array holders, ensuring consistent behavior.

* **Tests**
  * Introduced tests for string conversion, covering formatting, parsing, and error scenarios.

* **Chores**
  * Updated package version to 0.1.15.
  * Upgraded CI workflows and pre-commit hooks to use newer versions and improved Rust toolchain setup.

### CODE REVIEW

File: src/fixeduint/string_conversion.rs [hunk 1]
```diff
@@ -55,3 +56,49 @@ where
...
+        }
+
+        // 20 is sized for u64
+        let mut digit_blocks = [[0u8; MAX_DIGITS]; N];
+        let mut digit_count = 0;
+        let mut number = *self;
+        let ten = Self::from(10u8);
+
+        // Extract digits into our storage
+        while !number.is_zero() && digit_count < N * MAX_DIGITS {
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/fixeduint/string_conversion.rs:75]: That's acceptable

File: .github/workflows/coverage.yml [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
-# on: [ push, pull_request ]
 on: [ push ]
 
 jobs:
   build:
     runs-on: ubuntu-latest
 
     steps:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v4
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply .github/workflows/coverage.yml:10]: Nah i actually want the latest
